# Data wrangling and Subsetting

In [3]:
# import libraries
import pandas as pd
import numpy as np
import os

In [4]:
# creating path
path = r'C:\Users\bouas\instacart bascket analysis'

In [5]:
df_ords = pd.read_csv(os.path.join(path, 'data', 'original_data', 'orders.csv'))

In [6]:
df_prods = pd.read_csv(os.path.join(path, 'data', 'original_data', 'products.csv'), index_col = False)

In [46]:
# importing df-products
df_prods = pd.read_csv(os.path.join(path, 'data', 'original_data', 'products.csv'), index_col = False)

In [47]:
df_dep = pd.read_csv(os.path.join(path, 'data', 'original_data', 'departments.csv'), index_col = False)

In [48]:
# drop eval set
df_ords.drop(columns = ['eval_set'])

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
...,...,...,...,...,...,...
3421078,2266710,206209,10,5,18,29.0
3421079,1854736,206209,11,4,10,30.0
3421080,626363,206209,12,1,12,18.0
3421081,2977660,206209,13,1,12,7.0


In [49]:
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [50]:
# not displaying order id 
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [51]:
df_dep_t=df_dep.T

In [52]:
# new header
new_header = df_dep_t.iloc[0]

In [53]:
df_dep_t_1 = df_dep_t[1:]

In [86]:
#  new header
df_dep_t_1.columns = new_header

In [87]:
df_dep_t_1

department_id,department
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta
10,bulk


In [56]:
data_dict = df_dep_t_1.to_dict('index')

In [57]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

## 2/Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [80]:
# not displayeing Order number
df_ords_2['order_number'] = df_ords_2['order_number'].astype('str')

In [81]:
df_ords.dtypes

order_id                   object
user_id                    object
eval_set                   object
order_number               object
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [69]:
# changing the name order dow to orders day of the week
df_ords_2.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

## 3/Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [70]:
# changing the name to days since last order
df_ords_2.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

In [71]:
df_ords_2.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0


## 5/Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [15]:
# Here we should determine when ,so we need to have the count of the  orders hours of day 
df_ords_2['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

#### At 10 a.m., the highest number of orders are observed wich is the busiest hour for placing orders(288418)

## 5/Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary

In [16]:
# importing df-products
df_prods = pd.read_csv(os.path.join(path, 'data', 'original_data', 'products.csv'), index_col = False)

In [17]:
# using the fonction  : print(data_dict.get('4')) in our data dictionary 

### we will get {'department': 'produce'}

In [75]:
print(data_dict.get('4'))

{'department': 'produce'}


## 6/The sales team in your client’s organization wants to know more about breakfast item sales.

### searching in data dict: 14': {'department': 'breakfast'},

In [67]:
# processing to get item sales...
df_breakfast = df_prods.loc[df_prods['department_id'] == 14] 

In [68]:
df_breakfast

,product_id,product_name,aisle_id,department_id,prices
27,28,Wheat Chex Cereal,121,14,10.1
33,34,NaN,121,14,12.2
67,68,"Pancake Mix, Buttermilk",130,14,13.7
89,90,Smorz Cereal,121,14,3.9
210,211,Gluten Free Organic Cereal Coconut Maple Vanilla,130,14,3.6
...,...,...,...,...,...
49330,49326,Cereal Variety Fun Pack,121,14,9.1
49395,49391,Light and Fluffy Buttermilk Pancake Mix,130,14,2.0
49547,49543,Chocolate Cheerios Cereal,121,14,10.8
49637,49633,Shake 'N Pour Buttermilk Pancake Mix,130,14,14.2


## 7/items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [20]:
# dinner partie list form data dictionnary : 
# 1 frozen : to prepare rapidly  the dinner  so that s why it can be from dinner partie
# 12 meat seafood, 5 alkhool ,  20 deli , 7 beverages

In [21]:
df_dinner_parties= df_prods.loc[df_prods['department_id'].isin([1,12,5,20,7])]

In [76]:
df_dinner_parties

,product_id,product_name,aisle_id,department_id,prices
2,3,Robust Golden Unsweetened Oolong Tea,94,7,4.5
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1,10.5
6,7,Pure Coconut Water With Orange,98,7,4.4
7,8,Cut Russet Potatoes Steam N' Mash,116,1,1.1
9,10,Sparkling Orange Juice & Prickly Pear Beverage,115,7,8.4
...,...,...,...,...,...
49682,49678,Large Chicken & Cheese Taquitos,129,1,3.4
49685,49681,Spaghetti with Meatballs and Sauce Meal,38,1,6.9
49686,49682,California Limeade,98,7,4.3
49688,49684,"Vodka, Triple Distilled, Twist of Vanilla",124,5,5.3


## 8/It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

### 11657 rows × 5 columns

### 9/Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [72]:
df_USER1=df_ords_2.loc[df_ords_2['user_id'] == '1']

In [73]:
df_USER1

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
5,3367565,1,6,2,7,19.0
6,550135,1,7,1,9,20.0
7,3108588,1,8,1,14,14.0
8,2295261,1,9,1,16,0.0
9,2550362,1,10,4,8,30.0


##  10/You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [74]:
df_USER1.describe()

,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order
count,11.000000,11.000000,11.000000,10.000000
mean,6.000000,2.636364,10.090909,19.000000
std,3.316625,1.286291,3.477198,9.030811
min,1.000000,1.000000,7.000000,0.000000
25%,3.500000,1.500000,7.500000,14.250000
50%,6.000000,3.000000,8.000000,19.500000
75%,8.500000,4.000000,13.000000,26.250000
max,11.000000,4.000000,16.000000,30.000000


## 11/Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder

In [88]:
df_ords.to_csv(os.path.join(path, 'data','prepared_data', 'orders_wrangled.csv'))

## 12/Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [89]:
df_dep.to_csv(os.path.join(path, 'data','prepared_data', 'departments_wrangled.csv'))